p2p/net/swarm/swarm_dial.go

```diff
@@ -465,10 +465,21 @@ func (s *Swarm) dialAddr(ctx context.Context, p peer.ID, addr ma.Multiaddr) (con
 	return connC, nil
 }
 
+var ConnSetupTimeout = time.Minute * 5
+
 // dialConnSetup is the setup logic for a connection from the dial side. it
 // needs to add the Conn to the StreamSwarm, then run newConnSetup
 func dialConnSetup(ctx context.Context, s *Swarm, connC conn.Conn) (*Conn, error) {
 
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		deadline = time.Now().Add(ConnSetupTimeout)
+	}
+
+	if err := connC.SetDeadline(deadline); err != nil {
```

Comment:
Good point. Is `SetDeadline` the best way to solve it? Seems a thing that helps towards the solution, but it would be nice to have a way to tell when to close connections and with which criteria.